### DISCUSSION TITLE
V2: Notification Helper Methods?

### DISCUSSION THREAD

**Discussion by jon-sully** [excid3/noticed#360] [General]
Might be missing it, but as I work through some docs changes / updates I'm trying to trace the code-paths behind each chunk of docs. When talking about the Notifier helper methods (what I've written so far:)

---

##### Helper Methods

Notifiers can implement various helper methods that make it easier to render the resulting notification directly. While delivering by email (which has its own ActionMailer stack for rendering and processing) may or may not need these helpers, for example, rendering a notification in a web-view with ERB presents a case where these helpers are useful:

```erb
<div>
  <% @user.notifications.each do |notification| %>
    <%= link_to notification.message, notification.url %>
  <% end %>
</div>
```

---

I'm not finding code that supports this? I can't find anywhere that would proxy methods defined on a `Noticed::Event` sub-class over to a Notification instance (or rather, would delegate unknown method calls from a Notification instance to its `event`). Did that get dropped in V2 or am I totally missing it? 😅

**Comment by excid3** [comment]
A Notification belongs to an event (the event is the Notifier) now since we split it into 2 models. Event stores the params, Notification stores the recipients.

 To access methods on the Event, you have to use the association:

```erb
<div>
  <% @user.notifications.includes(:event).each do |notification| %>
    <%= link_to notification.event.message, notification.event.url %>
  <% end %>
</div>
```

**Reply by jon-sully** [reply]
Yeah, that makes sense. I appreciate the split of the data-model; I think it has a lot of really good benefits (hence the many closed issues in the repo now!) with only a little complexity cost.

This is an interesting case (I think, at least) though. While Notifiers _are_ Events, it seems like the Event records are abstracted away from the application / under-the-hood of Noticed. Pretty much everything about the application's integration with Noticed circles around the _notifications_ rather than the events (which I like). To keep up that abstraction, I do wonder if it's worth delegating missing methods over to the Event so that the application-level API remains `notification.message` / `notification.url` 🤔. What do you think?

**Reply by excid3** [reply]
It's possible we could do a `delegate_missing_to :event` in the Noticed::Notification to make that more "invisible" to the user. 

The main issue is that you no longer have access to `recipient` on those methods like in v1. A generic method like `url` is fine though.

**Reply by jon-sully** [reply]
Ah, good point. So there's really _two_ kind of "helper methods" for a Notifier then, the kind that are just essentially lambdas for `if:` / `unless:` hooks, which [_do_](https://github.com/excid3/noticed/blob/797878fe65049151ce5375d596dd25ac96eec8d1/lib/noticed/delivery_method.rb#L38) get executed in the context of the notification, and are intended just to yield a truthy or falsey value for that `if:`/`unless:`, and the kind that intended to be used from the notification side (helpers for rendering, for example, as above).

Not being able to access the recipient from that latter type of helper feels like it could be an issue for some. Technically this is a backwards-feeling approach that could work...

```ruby
class NewCommentNotifier < Noticed::Event
  # ...
  
  def message(recipient)
    "#{recipient.name}, you have a new comment to review"
  end
end
```

+

```erb
<div>
  <% @user.notifications.includes(:event).each do |notification| %>
    <%= link_to notification.event.message(notification.recipient)  %>
  <% end %>
</div>
```

But... that definitely feels gross.

Not that this option feels _that_ much cleaner, but I wonder if we could just slap an attribute on Event for `@current_notification` and override the getter on Notification to set it? Something like...

```ruby
class Notification < ApplicationRecord
  # ...
  delegate_missing_to :event

  def event
    super.tap { |e| e.current_notification = self }
  end
end

class Event < ApplicationRecord
  attr_writer :current_notification

  def recipient
    @current_notification&.recipient
  end
end
```

Would (I think) yield a usage like:

```ruby
class NewCommentNotifier < Noticed::Event
  # ...
  
  def message
    "#{recipient.name}, you have a new comment to review"
  end
end
```

+

```erb
<div>
  <% @user.notifications.includes(:event).each do |notification| %>
    <%= link_to notification.message  %>
  <% end %>
</div>
```

But it's not the prettiest chunk of code in the world..

**Reply by excid3** [reply]
That's pretty much where I landed too. I wasn't sold on it especially because it blurs the 2 models together and I want to make sure it's clear there is a record for the each recipient and a primary event that contains the shared data.

In this case, I would probably alter the wording so it's more clear:

```erb
<div>
  <% @user.notifications.includes(:event).each do |notification| %>
    <%= link_to notification.event.message_for(notification)  %>
  <% end %>
</div>
```

```ruby
class NewCommentNotifier < Noticed::Event
  # ...

  def message_for(notification)
    "#{notification.recipient.name}, you have a new comment to review"
  end
```

This is a bit more clear as the notifier is then responsible for deciding how a message is rendered for an individual notification. It isn't as seamless as `notification.message`, but it's not awful.

Since the Noticed::Notification is not STI, we don't have an easy way of adding methods to it dynamically like we do with Notifiers. Maybe Noticed::Notification could also be STI?

```ruby
class NewCommentNotifier < Noticed::Event
  class Notification < Noticed::Notification
    def message
      "#{recipient.name}, you have a new comment to review"
    end
  end
end
```

**Reply by jon-sully** [reply]
Class-in-a-class crossed my mind as well, though I didn't humor the thought because I figured it'd be gnarly 😆 but I just went down that road a bit and got towards this...

```ruby
# ~/app/notifiers/new_comment_notifier.rb

class NewCommentNotifier < Noticed::Event
  deliver_by :action_cable
  deliver_by :email, mailer: "CommentMailer", if:  ->(recipient) { !!recipient.preferences[:email] }

  bulk_deliver_by :discord do |config|
    config.url = "https://discord.com/xyz/xyz/123"
    config.json = -> {
      {
        message: message
      }
    }
  end

  module NotificationHelpers
    # I18n helpers
    def message
      t(".message")
    end

    # URL helpers are accessible in notifications
    # Don't forget to set your default_url_options so Rails knows how to generate urls
    def url
      post_path(params[:post])
    end
  end
end

# ---

class Notification < ApplicationRecord
  after_initialize do |notification|
    notification.extend(event.class::NotificationHelpers)
  end
end
```

The pain that lead me there is that at the User level (in our running example), the user will still just have-many notifications and when they pull `@user.notifications`, they should be enhanced with the methods defined in the specific Event-type.

Perhaps a road worth exploring? I feel like this isn't terrible 🤔

**Reply by excid3** [reply]
Without STI, you would be defining a single `message` method that would have to be shared across _all_ Notifiers which doesn't seem like it solves any problems.

I am really thinking the STI solution for Notifications may be ideal. We can pre-define that class and then you can override it to add methods.

**Reply by excid3** [reply]
Alright, well that turned out pretty slick with STI. We can dynamically define a `Notification` class inside each notifier, then like ActiveSupport::Concern, we can provide a block that gets evaluated in the `Notification` class to add methods.

```ruby
class NewCommentNotifier < Noticed::Event
  notification_methods do
    def message
      "hey its me"
    end
  end
end

NewCommentNotifier.deliver(User.first)
Noticed::Notification.last #=> NewCommentNotifier::Notification
Noticed::Notification.last.message #=> "hey its me"
```

**Reply by excid3** [reply]
This turned out super clean and it was really only 2 lines of code to implement. Merged into main. 🚀

**Reply by jon-sully** [reply]
Ahh that turned out really nice. Awesome! 💯

**Comment by jon-sully** [comment]
Just another general note — V2 removes callbacks for custom delivery methods. Neither here nor there, just pointing it out!

From V1:

---

#### Callbacks

Callbacks for delivery methods wrap the *actual* delivery of the notification. You can use `before_deliver`, `around_deliver` and `after_deliver` in your custom delivery methods.

```ruby
class DeliveryMethods::Discord < Noticed::DeliveryMethods::Base
  after_deliver do
    # Do whatever you want
  end
end
```

**Reply by excid3** [reply]
We can remove any mention of callbacks. I don't think they're terribly useful and I'd rather you implement a custom delivery method if you need something like that.

**Reply by ykamin-booth** [reply]
Can you add that callbacks are removed in the upgrade to v2 docs?
